### PR TITLE
Handling of negative `n` in hybrid::nth

### PR DESCRIPTION
**Pull request by romainfrancois**
Also, simplification of `lead` and `lag` following @krlmlr comments

### CODE REVIEW

File: inst/include/dplyr/hybrid/hybrid.h [hunk 1]
```diff
@@ -330,7 +344,11 @@ SEXP hybrid_do(SEXP expr, const SlicedTibble& data, const DataMask<SlicedTibble>
       int n;
 
       if (expression.is_unnamed(0) && expression.is_column(0, x) && expression.is_named(1, s_n) && expression.is_scalar_int(1, n)) {
-        return lag_1(data, x, n, op);
+        if (n > 0) {
+          return lag_1(data, x, n, op);
+        } else {
+          return echo(x.data, op);
```
Comment (krlmlr) [inline inst/include/dplyr/hybrid/hybrid.h]: Do we want to get rid of `echo()` ? If not we still need to check for `n == 0` here.

File: tests/testthat/test-hybrid.R [hunk 2]
```diff
@@ -191,6 +191,16 @@ test_that("nth(<column>, n = <int-ish>) is hybrid", {
   expect_not_hybrid(d, nth(chr, n = NA))
 })
 
+test_that("hybrid nth() handles negative n (#3821)", {
+  d <- tibble(int = 1:2, dbl = c(1,2), chr = c("a", "b"))
+  res <- summarise(d,
+    int = nth(int, !!(-1)),
```
Comment (krlmlr) [inline tests/testthat/test-hybrid.R]: Can we leave out the `!!()` with #3832? Perhaps we can have two tests, one with, one without?

File: inst/include/dplyr/hybrid/hybrid.h [hunk 3]
```diff
@@ -120,14 +130,14 @@ SEXP hybrid_do(SEXP expr, const SlicedTibble& data, const DataMask<SlicedTibble>
 
       Column x;
       if (expression.is_unnamed(0) && expression.is_column(0, x)) {
-        return first1_(data, x, op);
+        return nth2_(data, x, 1, op);
```
Comment (krlmlr) [inline inst/include/dplyr/hybrid/hybrid.h]: I don't mind keeping `first1_()` and friends, they can be simple forwarders to `nth2_()` . We have a lot of boilerplate in this function, seeing `s_first` map to `nth1_()` may seem odd. But it's not particularly important.

Comment (romainfrancois) [reply inst/include/dplyr/hybrid/hybrid.h]: Yeah i agree, i was getting too excited about deleting code.

File: tests/testthat/test-hybrid.R (hunk 2 above)
Comment (romainfrancois) [reply tests/testthat/test-hybrid.R]: Sure. We don’t need it here.

File: inst/include/dplyr/hybrid/hybrid.h (hunk 1 above)
Comment (romainfrancois) [reply inst/include/dplyr/hybrid/hybrid.h]: I kind of like it. Maybe i can stash it in lag_1 to reduce the noise in the hybrid_do dispatcher.

Comment (romainfrancois) [reply inst/include/dplyr/hybrid/hybrid.h]: ✅

**Pull request merged**